content/knowledge-base/guides/write-a-composition-function-in-python.md

```diff
@@ -0,0 +1,761 @@
+-->
+In v1.14 of the Crossplane CLI `crossplane beta xpkg init` just clones a
+template GitHub repository. In a future release the command will automate tasks
+like replacing the template name with the new function's name. See Crossplane
+issue [#4941](https://github.com/crossplane/crossplane/issues/4941) for details.
+<!-- vale gitlab.FutureTense = YES -->
+{{</hint>}}
+
+Edit `package/crossplane.yaml` to change the package's name before you start
+adding code. Name your package `function-xbuckets`.
+
+Some functions accept a configuration input. You configure the input when you
+write a Composition that uses the function. The `package/input` directory
+defines the OpenAPI schema for the a function's input. 
+
+The function in this guide doesn't accept an input. For this function you
+should delete the `package/input` directory. The
+[composition functions](https://docs.crossplane.io/latest/concepts/composition-functions)
+documentation explains more the input to a composition function.
+
+{{<hint "important">}}
+If you're writing a function that does use an input type, don't delete the
+`package/input` directory. Instead edit the file to be specific to your
+function.
+
+The kind `Input` is a placeholder value. The API group
+`template.fn.crossplane.io` is, too. Change the kind and API group to something
+meaningful to your function. Edit the `openAPIV3Schema` to represent your
+function's input schema.
+{{</hint>}}
+
+## Edit the template to add the function's logic
+
+You add your function's logic to the
+{{<hover label="hello-world" line="1">}}RunFunction{{</hover>}}
+method in `function/fn.py`. When you first open the file it contains a "hello
+world" function.
+
+```python {label="hello-world"}
+class FunctionRunner(grpcv1beta1.FunctionRunnerService):
+    def __init__(self):
+        self.log = logging.get_logger()
+
+    async def RunFunction(
+        self, req: fnv1beta1.RunFunctionRequest, _: grpc.aio.ServicerContext
+    ) -> fnv1beta1.RunFunctionResponse:
+        log = self.log.bind(tag=req.meta.tag)
+        log.info("Running function")
+
+        rsp = response.to(req)
+
+        example = ""
+        if "example" in req.input:
+            example = req.input["example"]
+
+        # TODO: Add your function logic here!
+        response.normal(rsp, f"I was run with input {example}!")
+        log.info("I was run!", input=example)
+
+        return rsp
+```
+
+All Python composition functions have a `RunFunction` method. Crossplane passes
+everything the function needs to run in a
+{{<hover label="hello-world" line="7">}}RunFunctionRequest{{</hover>}} object.
+The function tells Crossplane what resources it should compose by returning a
+{{<hover label="hello-world" line="22">}}RunFunctionResponse{{</hover>}} object.
+
+{{<hint "tip">}}
+Crossplane generates the `RunFunctionRequest` and `RunFunctionResponse` objects
+using [Protocol Buffers](https://protobuf.dev). You can find detailed schemas
```

Comment:
Vale was complaining about `protobuf`, despite it being part of a link URL. I tried putting the link on a line by itself without success. In the end I added it as allowed jargon, since I could imagine us wanting to say "protobuf" rather than "Protocol Buffers" outside of links too.

Reply:
Ah, I did actually use `protobuf` once outside a link down on line 357. I'm guessing it was complaining about that, but incorrectly telling me the error was coming from this line for some reason.

Reply:
I've seen that behavior before and it's a bug I haven't been able to chase down. I haven't been able to figure out what Vale is keying on to put the error in the wrong place.